### PR TITLE
CSV: Use placeholders when header and data length differs

### PR DESCRIPTION
**Pull request by yuferpegom**
This PR adds a couple of function that basically allows the user to include the empty fields (using some placeholders) in cases where there are more headers than data (or vice-versa) in the csv being transformed to a Map.

Examples:
 1. When there are more data than headers
```csv
eins,zwei
11,12,13
```
maps to
```scala
Map("eins" -> "11", "zwei" -> "12", "Missing header" -> "13")
```
The "Missing header" value is the default one, the user has the option to pass a custom value for this placeholder.

2. When there are more headers than data
```csv
eins,zwei,drei,vier,fünt
11,12,13
```
maps to
```scala
Map("eins" -> "11", "zwei" -> "12", "dreir" -> "13", "vier" -> "", "fünt" -> "")
```

This would be helpful, especially in the second case, when I want to keep the headers even when I don't have values associated with them.

### CODE REVIEW

**Comment by ennru** [comment]
Thank you for this suggestion.

What should happen if there is more than one extra data column?

**Comment by yuferpegom** [comment]
It should just add another header. If the user set a custom one it will add that one otherwise it will use the one configured by default:

1. Default placeholder:
```csv
eins,zwei
11,12,13,14
```
maps to
```scala
Map("eins" -> "11", "zwei" -> "12", "Missing header" -> "13" ,  "Missing Header" -> "14")
```

2. Custom placeholder
```csv
eins,zwei
11,12,13,14
```
maps to
```scala
Map("eins" -> "11", "zwei" -> "12", "custom" -> "13" ,  "custom" -> "14")
```

I think that this way is easier to understand what happened to the data (this is more helpful from the developer's point of view).

I also think that the more valuable use case for this change is when there are more headers than data as it is possible that the user wants to keep the data even when he might have forgotten to add a couple of commas on its input csv.

**Comment by ennru** [comment]
The Map won't be able to hold multiple values with the same key.

**Comment by yuferpegom** [comment]
Your right. So, It think that it can be bypassed by adding some character to the placeholder, like a number. Something like 

1. Default placeholder:
```csv
eins,zwei
11,12,13,14
```
maps to
```scala
Map("eins" -> "11", "zwei" -> "12", "Missing header" -> "13" ,  "Missing Header_1" -> "14")
```

What do you think? Any idea is also welcome, thanks!

**Comment by seglo** [comment]
> Your right. So, It think that it can be bypassed by adding some character to the placeholder, like a number.

Sounds reasonable to me. A 0-based index appended to the default missing key. Using your example:

```
CsvToMap.toMapCombineAll(
  missingHeaderDefault = "MissingHeader"
)
```

would return

```
Map("eins" -> "11", "zwei" -> "12", "MissingHeader0" -> "13" ,  "MissingHeader1" -> "14")
```

I would also suggest supporting a default value for missing values too.

```
CsvToMap.toMapCombineAll(
  missingValueDefault = "(missing)"
)
```

would return

```
Map("eins" -> "11", "zwei" -> "12", "dreir" -> "13", "vier" -> "(missing)", "fünt" -> "(missing)")
```

You'll need to support a javadsl as well.

**Comment by seglo** [comment]
@yuferpegom If you can follow up on this PR soon it can be included in the Alpakka `3.0.0-M1` release soon.

**Comment by yuferpegom** [comment]
Oh I will, thanks

**Comment by lightbend-cla-validator** [comment]
Hi @yupegom,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[https://www.lightbend.com/contribute/cla](https://www.lightbend.com/contribute/cla)

**Review by seglo — COMMENTED** [review]
I think this is almost there. Just a few small things.

File: csv/src/test/scala/docs/scaladsl/CsvToMapSpec.scala [hunk 1]
```diff
@@ -189,5 +195,278 @@ class CsvToMapSpec extends CsvSpec {
       // #column-names
     }
 
+    "parse header and decode data line. Be OK with more headers column than data (including the header in the result)" in assertAllStagesStopped {
+      // #header-line
+      import akka.stream.alpakka.csv.scaladsl.{CsvParsing, CsvToMap}
+
+      // #header-line
+      val future =
+        // format: off
```
[the remark this answers is not in the document]
Comment (yuferpegom) [reply csv/src/test/scala/docs/scaladsl/CsvToMapSpec.scala:204]: This allows keeping the special indentation to represent the columns and rows being passed as params. 

BTW, I'm just following what was already done before in the spec.

**Comment by yuferpegom** [comment]
@seglo I have addressed your last comments, please take a look when you have a chance and than you!

**Review by ennru — APPROVED** [review]
LGTM.

**Pull request merged**